### PR TITLE
storage: Make snapshot trait return iterator instead of cursor

### PR DESCRIPTION
**Pull request by Connor1996**
### What problem does this PR solve?

A prerequisite change of #9758, so TTLIterator can wrap RegionIterator directly

### What is changed and how it works?

Change iter() of Snapshot trait from
```rust
     fn iter(&self, iter_opt: IterOptions, mode: ScanMode) -> EngineResult<Cursor<Self::Iter>> 
```
to 
```rust
     fn iter(&self, iter_opt: IterOptions) -> EngineResult<Self::Iter> 
```

### Check List 

Tests 

- No code

### Release note 

- None

### CODE REVIEW

File: src/server/raftkv.rs [hunk 1]
```diff
@@ -542,33 +542,18 @@ impl<S: Snapshot> EngineSnapshot for RegionSnapshot<S> {
...
         fail_point!("raftkv_snapshot_iter_cf", |_| Err(box_err!(
             "injected error for iter_cf"
         )));
-        let prefix_seek = iter_opt.prefix_seek_used();
-        Ok(Cursor::new(
-            RegionSnapshot::iter_cf(self, cf, iter_opt)?,
-            mode,
-            prefix_seek,
-        ))
+        Ok(RegionSnapshot::iter_cf(self, cf, iter_opt)?)
```
Comment (Little-Wallace) [inline src/server/raftkv.rs:556]: How about refactoring it with 
```
RegionSnapshot::iter_cf(self, cf, iter_opt).map_err(kv::Error::from)
```

**Review by Little-Wallace — COMMENTED** [review]
rest LGTM

**Review by Little-Wallace — APPROVED** [review]
LGTM

File: src/storage/kv/cursor.rs [hunk 2]
```diff
@@ -552,11 +568,18 @@ impl<'a, S: 'a + Snapshot> CursorBuilder<'a, S> {
         if let Some(ts) = self.hint_max_ts {
             iter_opt.set_hint_max_ts(Bound::Included(ts.into_inner()));
         }
+        iter_opt.set_key_only(self.key_only);
```
[the remark this answers is not in the document]
Comment (Connor1996) [reply src/storage/kv/cursor.rs:571]: the signature of `set_key_only` isn't `mut self` but `&mut self`

[the remark this answers is not in the document]
Comment (BusyJay) [reply src/storage/kv/cursor.rs:571]: So all methods use `mut self` except `set_key_only`?

[the remark this answers is not in the document]
Comment (Connor1996) [reply src/storage/kv/cursor.rs:571]: only `set_max_skippable_internal_keys`, `use_prefix_seek` and `set_prefix_same_as_start` are `mut self`, others are `&mut self`😂

[the remark this answers is not in the document]
Comment (BusyJay) [reply src/storage/kv/cursor.rs:571]: Better keep them consistent to avoid missing updates.

**Review by BusyJay — APPROVED** [review]

**Pull request merged**